File: backend/internal/models/fact.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type Fact struct {
	ID          primitive.ObjectID `bson:"_id,omitempty" json:"id"`
	Content     string            `bson:"content" json:"content"`
	Category    string            `bson:"category" json:"category"`
	Source      string            `bson:"source" json:"source"`
	Tags        []string          `bson:"tags" json:"tags"`
	Verified    bool              `bson:"verified" json:"verified"`
	CreatedAt   time.Time         `bson:"created_at" json:"created_at"`
	UpdatedAt   time.Time         `bson:"updated_at" json:"updated_at"`
	RelatedURLs []string          `bson:"related_urls" json:"related_urls"`
	Metadata    FactMetadata      `bson:"metadata" json:"metadata"`
}

type FactMetadata struct {
	Language    string   `bson:"language" json:"language"`
	Difficulty  string   `bson:"difficulty" json:"difficulty"`
	References  []string `bson:"references" json:"references"`
	Keywords    []string `bson:"keywords" json:"keywords"`
	Popularity  int      `bson:"popularity" json:"popularity"`
	LastServed  time.Time `bson:"last_served" json:"last_served"`
	ServeCount  int      `bson:"serve_count" json:"serve_count"`
}

type FactQuery struct {
	Category    string    `json:"category"`
	Tags        []string  `json:"tags"`
	StartDate   time.Time `json:"start_date"`
	EndDate     time.Time `json:"end_date"`
	SearchTerm  string    `json:"search_term"`
	Difficulty  string    `json:"difficulty"`
	Language    string    `json:"language"`
	Verified    *bool     `json:"verified"`
	Limit       int       `json:"limit"`
	Offset      int       `json:"offset"`
}

File: One Fact/One Fact/Models/Fact.swift
import Foundation

struct RelatedArticle: Codable, Identifiable {
    let id: Int
    let title: String
    let url: String
    let source: String
    let snippet: String
}

struct Fact: Codable, Identifiable {
    let id: String
    let content: String
    let category: String
    let source: String
    let tags: [String]
    let verified: Bool
    let createdAt: Date
    let updatedAt: Date
    let relatedURLs: [String]
    let metadata: FactMetadata
    
    var displayDate: Date {
        // Use createdAt for display purposes
        createdAt
    }
    
    var relatedArticles: [RelatedArticle] {
        // Convert relatedURLs to RelatedArticle objects
        return relatedURLs.enumerated().map { index, url in
            let keyword = (metadata.keywords?.first ?? category)
            let snippet = "Learn more about \(keyword) from \(source)"
            return RelatedArticle(
                id: index,
                title: "Related Article \(index + 1)",
                url: url,
                source: source,
                snippet: snippet
            )
        }
    }
    
    var url: String? {
        // Return the first related URL if available
        relatedURLs.first
    }
    
    enum CodingKeys: String, CodingKey {
        case id
        case content
        case category
        case source
        case tags
        case verified
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case relatedURLs = "related_urls"
        case metadata
    }
}

struct FactMetadata: Codable {
    let language: String?
    let difficulty: String?
    let references: [String]?
    let keywords: [String]?
    let popularity: Int
    let lastServed: Date
    let serveCount: Int
    
    enum CodingKeys: String, CodingKey {
        case language
        case difficulty
        case references
        case keywords
        case popularity
        case lastServed = "last_served"
        case serveCount = "serve_count"
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        language = try container.decodeIfPresent(String.self, forKey: .language) ?? ""
        difficulty = try container.decodeIfPresent(String.self, forKey: .difficulty) ?? ""
        references = try container.decodeIfPresent([String].self, forKey: .references) ?? []
        keywords = try container.decodeIfPresent([String].self, forKey: .keywords) ?? []
        popularity = try container.decodeIfPresent(Int.self, forKey: .popularity) ?? 0
        lastServed = try container.decode(Date.self, forKey: .lastServed)
        serveCount = try container.decodeIfPresent(Int.self, forKey: .serveCount) ?? 0
    }
}

struct APIResponse<T: Codable>: Codable {
    let success: Bool
    let data: T
}
